pkgs/development/compilers/llvm/15/compiler-rt/default.nix

```diff
@@ -0,0 +1,148 @@
+}:
+
+let
+
+  useLLVM = stdenv.hostPlatform.useLLVM or false;
+  bareMetal = stdenv.hostPlatform.parsed.kernel.name == "none";
+  haveLibc = stdenv.cc.libc != null;
+  inherit (stdenv.hostPlatform) isMusl;
+
+  baseName = "compiler-rt";
+
+  src = runCommand "${baseName}-src-${version}" {} ''
+    mkdir -p "$out"
+    cp -r ${monorepoSrc}/cmake "$out"
+    cp -r ${monorepoSrc}/${baseName} "$out"
+  '';
+in
+
+stdenv.mkDerivation {
+  pname = baseName + lib.optionalString (haveLibc) "-libc";
+  inherit version;
+
+  inherit src;
+  sourceRoot = "${src.name}/${baseName}";
+
+  nativeBuildInputs = [ cmake ninja python3 libllvm.dev ];
+  buildInputs = lib.optional stdenv.hostPlatform.isDarwin libcxxabi;
+
+  NIX_CFLAGS_COMPILE = [
+    "-DSCUDO_DEFAULT_OPTIONS=DeleteSizeMismatch=0:DeallocationTypeMismatch=0"
+  ];
+
+  cmakeFlags = [
+    "-DCOMPILER_RT_DEFAULT_TARGET_ONLY=ON"
+    "-DCMAKE_C_COMPILER_TARGET=${stdenv.hostPlatform.config}"
+    "-DCMAKE_ASM_COMPILER_TARGET=${stdenv.hostPlatform.config}"
+  ] ++ lib.optionals (useLLVM || bareMetal || isMusl) [
+    "-DCOMPILER_RT_BUILD_SANITIZERS=OFF"
+    "-DCOMPILER_RT_BUILD_XRAY=OFF"
+    "-DCOMPILER_RT_BUILD_LIBFUZZER=OFF"
+    "-DCOMPILER_RT_BUILD_PROFILE=OFF"
+    "-DCOMPILER_RT_BUILD_MEMPROF=OFF"
+    "-DCOMPILER_RT_BUILD_ORC=OFF" # may be possible to build with musl if necessary
+  ] ++ lib.optionals ((useLLVM && !haveLibc) || bareMetal) [
+    "-DCMAKE_C_COMPILER_WORKS=ON"
+    "-DCMAKE_CXX_COMPILER_WORKS=ON"
+    "-DCOMPILER_RT_BAREMETAL_BUILD=ON"
+    "-DCMAKE_SIZEOF_VOID_P=${toString (stdenv.hostPlatform.parsed.cpu.bits / 8)}"
+  ] ++ lib.optionals (useLLVM && !haveLibc) [
+    "-DCMAKE_C_FLAGS=-nodefaultlibs"
+  ] ++ lib.optionals (useLLVM) [
+    "-DCOMPILER_RT_BUILD_BUILTINS=ON"
+    #https://stackoverflow.com/questions/53633705/cmake-the-c-compiler-is-not-able-to-compile-a-simple-test-program
+    "-DCMAKE_TRY_COMPILE_TARGET_TYPE=STATIC_LIBRARY"
+  ] ++ lib.optionals (bareMetal) [
+    "-DCOMPILER_RT_OS_DIR=baremetal"
+  ] ++ lib.optionals (stdenv.hostPlatform.isDarwin) [
+    "-DDARWIN_macosx_OVERRIDE_SDK_VERSION=ON"
+    "-DDARWIN_osx_ARCHS=${stdenv.hostPlatform.darwinArch}"
+    "-DDARWIN_osx_BUILTIN_ARCHS=${stdenv.hostPlatform.darwinArch}"
+
+    # `COMPILER_RT_DEFAULT_TARGET_ONLY` does not apply to Darwin:
+    # https://github.com/llvm/llvm-project/blob/27ef42bec80b6c010b7b3729ed0528619521a690/compiler-rt/cmake/base-config-ix.cmake#L153
+    "-DCOMPILER_RT_ENABLE_IOS=OFF"
+  ];
+
+  outputs = [ "out" "dev" ];
+
+  patches = [
+    ./X86-support-extension.patch # Add support for i486 i586 i686 by reusing i386 config
+    ./gnu-install-dirs.patch
+    # ld-wrapper dislikes `-rpath-link //nix/store`, so we normalize away the
+    # extra `/`.
+    ./normalize-var.patch
+  ] # Prevent a compilation error on darwin
+    ++ lib.optional stdenv.hostPlatform.isDarwin ./darwin-targetconditionals.patch
+    ++ lib.optional stdenv.hostPlatform.isAarch32 ./armv7l.patch
```

Comment:
This patch is not applying and not needed:

```
compiler-rt-libc> unpacking sources
compiler-rt-libc> unpacking source archive /nix/store/nrg5h6yf18wfvllg9qywnhw98gay3ill-compiler-rt-src-15.0.2
compiler-rt-libc> source root is compiler-rt-src-15.0.2/compiler-rt
compiler-rt-libc> patching sources
compiler-rt-libc> applying patch /nix/store/qvyl9jgm1a1mlix44qbsgrd1ipwciamk-X86-support-extension.patch
compiler-rt-libc> patching file lib/builtins/CMakeLists.txt
compiler-rt-libc> Hunk #1 succeeded at 355 (offset 7 lines).
compiler-rt-libc> Hunk #2 succeeded at 735 with fuzz 2 (offset 8 lines).
compiler-rt-libc> applying patch /nix/store/8z4wq03gvjdyvpldnrl7x1sy14qfk1pk-gnu-install-dirs.patch
compiler-rt-libc> patching file cmake/base-config-ix.cmake
compiler-rt-libc> applying patch /nix/store/x0137mdpq76zfp2xvn3ncawys2g4c93i-normalize-var.patch
compiler-rt-libc> patching file cmake/Modules/CompilerRTUtils.cmake
compiler-rt-libc> Hunk #1 succeeded at 392 with fuzz 2 (offset 90 lines).
compiler-rt-libc> applying patch /nix/store/jd7mnr62kxgks1rwnrrfg22vwnggkpxf-armv7l.patch
compiler-rt-libc> patching file cmake/builtin-config-ix.cmake
compiler-rt-libc> Hunk #1 FAILED at 24.
compiler-rt-libc> 1 out of 1 hunk FAILED -- saving rejects to file cmake/builtin-config-ix.cmake.rej
compiler-rt-libc> patching file lib/builtins/CMakeLists.txt
compiler-rt-libc> Hunk #1 succeeded at 569 with fuzz 2 (offset 95 lines).
compiler-rt-libc> Hunk #2 FAILED at 596.
compiler-rt-libc> 1 out of 2 hunks FAILED -- saving rejects to file lib/builtins/CMakeLists.txt.rej
```

Reply:
I think I'm missing some context here; I know that you dropped this patch from `llvmPackages_14` (https://github.com/NixOS/nixpkgs/pull/178328) but I'm fuzzy on what exactly was causing the error originally (#87419) and what changed in upstream to make this no longer required.

Reply:
I think updating this patch to apply is just a matter of tweaking a few line numbers but since this patch has already been dropped from `llvmPackages_14.compiler-rt` I'll remove it from here as well.

I intend to replace `llvmPackages_git` with a copy of `llvmPackages_15` in a follow up PR so I'm going to hold off on mirroring this change in `llvmPackages_git` in this PR.

Reply:
I fixed those patches and reinstated them. Please keep.

Reply:
@Ericson2314 Can your point me to where you updated this patch? I don't see any pending or merged PRs on `nixpkgs` that reference the `armv7l` patch and I was not able to find anything pending in upstream either.

Also I think you may have more context than I do here; do you if/why this patch is still required?

Reply:
See also https://github.com/NixOS/nixpkgs/pull/150022#issuecomment-990825227

Reply:
https://github.com/NixOS/nixpkgs/pull/148602/files and d6636a5de6bb7d261adc2957c4a909ae035aae3b

Reply:
@Ericson2314 thanks for the links; I'll update the patch in this PR to match.

Do you have any information about how this bug manifests? @misuzu and I were able to build clang and compiler-rt on armv7 without this patch; do we need to actually use symbols from `libclang-rt` to see the error?

I'd like to add a regression test for this so we catch this sooner in the future.

Reply:
@rrbutani It shows up when compiling for weird arms. Like `arm6m` or `arm8m` if I am not mistaken.

Reply:
The patch only adds `armv7l` to several lists, would that really help compiling for some other weird architectures?

Reply:
This patch is still broken:
```
error: builder for '/nix/store/vq6bg0vldgfkvz8jsxakc8icnww7vav6-compiler-rt-libc-15.0.6.drv' failed with exit code 1;
       last 10 log lines:
       > patching file cmake/Modules/CompilerRTUtils.cmake
       > Hunk #1 succeeded at 392 with fuzz 2 (offset 90 lines).
       > applying patch /nix/store/jd7mnr62kxgks1rwnrrfg22vwnggkpxf-armv7l.patch
       > patching file cmake/builtin-config-ix.cmake
       > Hunk #1 FAILED at 24.
       > 1 out of 1 hunk FAILED -- saving rejects to file cmake/builtin-config-ix.cmake.rej
       > patching file lib/builtins/CMakeLists.txt
       > Hunk #1 succeeded at 569 with fuzz 2 (offset 95 lines).
       > Hunk #2 FAILED at 596.
       > 1 out of 2 hunks FAILED -- saving rejects to file lib/builtins/CMakeLists.txt.rej
       For full logs, run 'nix log /nix/store/vq6bg0vldgfkvz8jsxakc8icnww7vav6-compiler-rt-libc-15.0.6.drv'.
```

Reply:
Apologies; this should be fixed in 735660541acd064081c6dc9c7e6ef99ab3cd6846; @misuzu can you please try again?

Reply:
It builds now, thanks!